### PR TITLE
In Jacobian, re-use partial einsums (closes #168)

### LINKED ISSUE #168: Remove double loop in J
It can be changed into a single loop (thanks @darcey).

### PR DESCRIPTION
**Pull request by grantexley**

### CODE REVIEW

**Comment by davidweichiang** [comment]
- [ ] make selectable by some kind of optional argument to `sum_product`
- [ ] benchmarking
- [ ]

**Comment by davidweichiang** [comment]
The optional argument might relate to #170

**Comment by davidweichiang** [comment]
Could https://pypi.org/project/opt-einsum/ be used to find an ordering?

**Comment by grantexley** [comment]
image
image

**Comment by davidweichiang** [comment]
In the above graphs, "without -j" is the original code and "with -j" is the new code.

It would be great if "with -j" was always better, or at least only a little bit worse, so that we could just replace the old Jacobian with the new one.

File: fggs/sum_product.py [hunk 1]
```diff
@@ -149,6 +148,112 @@ def J(fgg: FGG, x: MultiTensor, inputs: MultiTensor, semiring: Semiring,
...
+    """The Jacobian of F. New version that takes out the double loop over edges"""
+    Jx = MultiTensor(x.shapes+x.shapes, semiring)
+    for n in x.shapes[0]:
+        for rule in fgg.rules(n):
+            edges = list(rule.rhs.edges())
+            if len(edges) == 1:
+                suffix_products = [sum_product_edges(fgg, rule.rhs.nodes(), set(), rule.rhs.ext + edges[0].nodes, x, inputs, semiring=semiring)]
+            else:
+                prefix_products, prefix_output_nodes = compute_products(fgg, (x, inputs), edges, rule, semiring, "forward")
+                suffix_products, suffix_output_nodes = compute_products(fgg, (x, inputs), edges, rule, semiring, "backward")
```
Comment (davidweichiang) [inline fggs/sum_product.py:162]: Would it be simpler just to reverse the edges and reverse the return values?

**Review by davidweichiang — APPROVED** [review]
I think it looks okay, though my memory is hazy on some of the details!

**Pull request merged**